modules/human-web-lite/sources/patterns-updater.es

```diff
@@ -0,0 +1,282 @@
+/*!
+ * Copyright (c) 2014-present Cliqz GmbH. All rights reserved.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+import logger from './logger';
+
+const SECOND = 1000;
+const MINUTE = 60 * SECOND;
+const HOUR = 60 * MINUTE;
+
+/**
+ * If you need to introduce incompatible changes to the the state
+ * persistence, you can bump this number to clear the persisted cache.
+ * It will be as if you start with a fresh installation of the extension.
+ *
+ * (If you are not sure whether the change is incompatible, it is
+ * a good idea to be conservative and bump this number anyway; it will
+ * only introduce little overhead.)
+ */
+const DB_VERSION = 1;
+
+function randBetween(min, max) {
+  return min + Math.random() * (max - min);
+}
+
+function clamp({ min, max, value }) {
+  return Math.min(Math.max(min, value), max);
+}
+
+/**
+ * Responsible for keeping the patterns up-to-date by polling
+ * the backend for changes. By design, it supports situations where
+ * the extension is frequently restarted; in other words, it is
+ * designed to work in an empheral service worker model and
+ * does not required a persistent background.
+ *
+ * The current implementation assumes that the "update" function
+ * gets triggered frequently enough to check for new patterns
+ * within the configured intervals.
+ *
+ * (If the need to trigger the "update" function becomes a problem,
+ * it could be lifted by using alternative browser APIs such as
+ * as the "Alarms API". But as calling update is a fast operation
+ * and the results are cached, it is safe to call it before
+ * each access to the patterns.)
+ */
+export default class PatternsUpdater {
```

Comment:
Naive question: cannot a ResourceLoader be used here?

Reply:
> Naive question: cannot a ResourceLoader be used here?

In the Manifest V3 context, background jobs with setInterval are not ideal IMHO (if we would to use the same implementation). That is why lazily updates, but with the guarantee that the resource is up-to-date when you need it (unless the network is down). Laziness also safes bandwidth if no updates are needed.

Reply:
got it, make sense!